File: 03_CustomView/app/src/main/java/com/example/tengfei/customview/view/CircleView.java
package com.example.tengfei.customview.view;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.util.AttributeSet;
import android.view.View;

/**
 * @author tengfei
 * date 2019/1/17 3:02 PM
 * email tengfeigo@outlook.com
 * description 自定义颜色可变的圆形控件，用于 BouquetLoading 中
 */
public class CircleView extends View {

    private Paint mPaint;

    public CircleView(Context context) {
        this(context, null);
    }

    public CircleView(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public CircleView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        mPaint = new Paint();
        mPaint.setAntiAlias(true);
        mPaint.setDither(true);
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);
        int cx = getWidth() / 2;
        int cy = getHeight() / 2;
        canvas.drawCircle(cx, cy, cx, mPaint);
    }

    public void exchangeColor(int color) {
        mPaint.setColor(color);
        invalidate();
    }
}

File: SpecialEffects/app/src/main/java/com/example/specialeffects/view/CircleView.kt
package com.example.specialeffects.view

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.util.AttributeSet
import android.view.View
import com.example.specialeffects.utils.px2dp

/**
 * @author tengfei
 * date 2019-10-08 21:47
 * email tengfeigo@outlook.com
 * description
 */
class CircleView
@JvmOverloads constructor(
    context: Context,
    attributeSet: AttributeSet? = null,
    defStyleAttr: Int = 0
) : View(context, attributeSet, defStyleAttr) {

    private var radious = px2dp(50F)
        get() = field
        set(value) {
            field = value
            invalidate()
        }

    var paint: Paint = Paint(Paint.ANTI_ALIAS_FLAG)

    init {
        paint.color = Color.RED
    }


    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        canvas.drawCircle((width / 2).toFloat(), (height / 2).toFloat(), radious, paint)
    }
}